sdk/trace/batch_span_processor.go

```diff
@@ -154,90 +158,76 @@ func WithBlocking() BatchSpanProcessorOption {
 	}
 }
 
+// exportSpans is a subroutine of processing and draining the queue.
+func (bsp *BatchSpanProcessor) exportSpans() {
+	bsp.timer.Reset(bsp.o.ScheduledDelayMillis)
+
+	if len(bsp.batch) > 0 {
+		bsp.e.ExportSpans(context.Background(), bsp.batch)
+		bsp.batch = bsp.batch[:0]
+	}
+}
+
 // processQueue removes spans from the `queue` channel until processor
 // is shut down. It calls the exporter in batches of up to MaxExportBatchSize
 // waiting up to ScheduledDelayMillis to form a batch.
 func (bsp *BatchSpanProcessor) processQueue() {
-	timer := time.NewTimer(bsp.o.ScheduledDelayMillis)
-	defer timer.Stop()
-
-	batch := make([]*export.SpanData, 0, bsp.o.MaxExportBatchSize)
+	defer bsp.stopWait.Done()
+	defer bsp.timer.Stop()
 
-	exportSpans := func() {
-		timer.Reset(bsp.o.ScheduledDelayMillis)
-
-		if len(batch) > 0 {
-			bsp.e.ExportSpans(context.Background(), batch)
-			batch = batch[:0]
-		}
-	}
-
-loop:
 	for {
 		select {
 		case <-bsp.stopCh:
-			break loop
-		case <-timer.C:
-			exportSpans()
+			return
+		case <-bsp.timer.C:
+			bsp.exportSpans()
 		case sd := <-bsp.queue:
-			batch = append(batch, sd)
-			if len(batch) == bsp.o.MaxExportBatchSize {
-				if !timer.Stop() {
-					<-timer.C
+			bsp.batch = append(bsp.batch, sd)
+			if len(bsp.batch) == bsp.o.MaxExportBatchSize {
+				if !bsp.timer.Stop() {
+					<-bsp.timer.C
 				}
-				exportSpans()
-			}
-		}
-	}
-
-	for {
-		select {
-		case sd := <-bsp.queue:
-			if sd == nil { // queue is closed
-				go throwAwayFutureSends(bsp.queue)
-				exportSpans()
-				return
+				bsp.exportSpans()
 			}
-
-			batch = append(batch, sd)
-			if len(batch) == bsp.o.MaxExportBatchSize {
-				exportSpans()
-			}
-		default:
-			// Send nil instead of closing to prevent "send on closed channel".
-			bsp.queue <- nil
 		}
 	}
 }
 
-func throwAwayFutureSends(ch <-chan *export.SpanData) {
-	for {
-		select {
-		case <-ch:
-		case <-time.After(time.Minute):
-			return
+// drainQueue awaits the any caller that had added to bsp.stopWait
+// to finish the enqueue, then exports the final batch.
+func (bsp *BatchSpanProcessor) drainQueue() {
+	for sd := range bsp.queue {
+		bsp.batch = append(bsp.batch, sd)
+		if len(bsp.batch) == bsp.o.MaxExportBatchSize {
+			bsp.exportSpans()
 		}
 	}
+	bsp.exportSpans()
 }
 
 func (bsp *BatchSpanProcessor) enqueue(sd *export.SpanData) {
 	if !sd.SpanContext.IsSampled() {
 		return
 	}
 
-	select {
-	case <-bsp.stopCh:
-		return
-	default:
-	}
+	// This ensures the bsp.queue<- below does not panic as the
+	// processor shuts down.
+	defer func() {
+		_ = recover()
```

Comment:
This could be 

```
	defer func() {
		x := recover()
		switch x := x.(type) {
		case nil:
			return
		case runtime.Error:
			if x.Error() == "send on closed channel" {
				return
			}
		}
		panic(x)
	}()
```

so this does not catch all the panic and make debugging easier. Performance should be the same.

Reply:
Done.